[WIP] Add Moon+ Reader support

stbenjam opened the pull request:
Note: hold off on reviewing this more until some of the separate features I've extracted from this are merged:
- [ ] File chooser, https://github.com/davidfor/calibre-annotations/pull/25
- [ ] Default and HTML colors, https://github.com/davidfor/calibre-annotations/pull/24

--------

This adds support for importing annotations from a Moon+ Reader backup
file (.mrpro).  

This extracts the annotations from there. Feedback welcome, my python is pretty out of date.

Screenshots:

![moon1](https://user-images.githubusercontent.com/429763/163681359-093f3ab1-2e49-4824-a316-92b1db78a40c.png)

![moon2](https://user-images.githubusercontent.com/429763/163681362-1e36b7fe-19b3-4e7b-85fd-64e12c9a7ede.png)

![moon3](https://user-images.githubusercontent.com/429763/163681365-550083d7-acea-486c-86f3-72d0bcf2c17c.png)

![moon5](https://user-images.githubusercontent.com/429763/163681580-e9f51936-b19e-4b36-b1d3-1e59aba319e3.png)

Fixes #19

stbenjam committed:
Add Moon+ Reader support

This adds support for importing annotations from a Moon+ Reader backup
file (.mrpro).

.mrpro is just a zip file, that contains a SQLite3 databse inside. This
extracts the annotations from there, and imports them with the
Annotations plugin.

stbenjam commented on readers/MoonReaderApp.py:
```diff
@@ -0,0 +1,186 @@
+            ann_mi.annotation_id = row[1]
+            # FIXME - colors in the moon reader db are integer values reflecting RGBA. I guess I'd need to figure out
+            #   an approximation of which color is being used. For now, everything's Yellow...
+            ann_mi.highlight_color = 'Yellow'
```
Just to clarify: there's a limited set of colors I can use, right? I can't just put in an HTML hex color or something here?

stbenjam replied:
N/m, I got it working

stbenjam committed:
Support direct use of HTML colors, and other clean-ups

stbenjam committed:
Make a better guess about backup filename

stbenjam committed:
Allow drag-and-drop filename, and don't clear result when selecting default value

stbenjam commented on common_utils.py:
```diff
@@ -221,21 +221,24 @@ def dragMoveEvent(self, event):
+                    u = unicode(raw, 'utf-8')
+                self.setPlainText(u)
+            else:
+                self.setPlainText(path)
```
This allows a user to drag-and-drop a binary file, and the field will contain the path instead of contents.

aik099 replied:
Drag-and-drop file into where? Some supporting screenshots and explanations are needed.

---

This sounds like a solid feature, which should go into a separate PR including a documentation update.

stbenjam replied:
The exporting readers already show a dialog, which you can drag and drop a file into.  For the other exporting readers, it appears to be text content. There's a screen shot of this dialog in my description.

Moon has binary content, so in this case it puts the path instead of the file content.  This seems only relevant to Moon's binary backup, so does this really need to be a separate PR?

stbenjam replied:
I think the better approach here is to let `ExportingReader`s have the option to expose a file chooser, instead of text box.

I'll put together something as a separate PR.

aik099 replied:
Looking forward to it.

stbenjam commented on common_utils.py:
```diff
@@ -221,21 +221,24 @@ def dragMoveEvent(self, event):
-        url = str(data.data(mime))
-        path = urlparse(url).path.strip()
-        scheme = urlparse(url).scheme
+        url = urlparse(str(data.data(mime), encoding='utf-8').rstrip())
+        path = url.path.strip()
+        scheme = url.scheme
```
In my testing, if I didn't encode it to utf-8, the urlparser wasn't able to parse a file path

stbenjam commented on common_utils.py:
```diff
@@ -365,9 +368,6 @@ def import_annotations_dialog_clicked(self, button):
         if self.dialogButtonBox.buttonRole(button) == QDialogButtonBox.AcceptRole:
-            # Remove initial_dialog_text if user clicks OK without dropping file
-            if self.text() == self.rac.initial_dialog_text:
-                self.pte.clear()
```
Moon Reader tries to make a guess at the file path, since it always picks `<date>.mrpro` -- with this code, it doesn't let a user accept the default value.

aik099 replied:
What are the values for these function result and the property:

* `self.text()`
* `self.rac.initial_dialog_text` 

Not sure what removed code actually does for other reader classes. We surely need to understand that before removing it.

stbenjam replied:
self.text() contains the contents of the exporting reader dialog window.  The initial dialog tests for the other exporting readers are already blank:

https://github.com/davidfor/calibre-annotations/blob/8f0104a35f2919c1780e0758266159febc891e21/readers/_BluefireReader.py#L88

https://github.com/davidfor/calibre-annotations/blob/8f0104a35f2919c1780e0758266159febc891e21/readers/GoodReader.py#L82

Except for Marvin, which just tells the user to drag and drop a file.

Clearing the field if the text didn't change doesn't really make much sense to me, and I don't think it breaks anyone.

aik099 replied:
I see. Still better to do this in separate PR, since it's not breaking Moon Reader integration.

aik099 commented on annotations.py:
```diff
@@ -190,11 +192,18 @@ def to_HTML(self, header=''):
+                    elif htmlColorPattern.match(agroup.highlightcolor):
+                        msg = "Found valid RGB(A) HTML color '%s', using" % agroup.highlightcolor
+                        dt_bgcolor = agroup.highlightcolor
+                        dt_fgcolor = COLOR_MAP['Default']['fg']
```
Not all background colors look well with the black foreground color. Therefore it's better to add support for specifying a custom foreground color (optional setting, that defaults to `COLOR_MAP['Default']['fg']` as now). 

---

Changes in this file look like a nice feature and should go into a standalone PR with an explanation of how to use that feature. Probably sample annotation importer file also should be updated with an examples.

stbenjam replied:
In the case of moon, there's no foreground color so where am I supposed to set this, a plugin-wide setting?

stbenjam replied:
I suppose it could go under "Modify appearance" to set a default highlight foreground/background? I can put together a separate PR for this.

stbenjam replied:
#24

aik099 replied:
#24 reviewed.

aik099 commented on common_utils.py:
```diff
@@ -221,21 +221,24 @@ def dragMoveEvent(self, event):
-        url = str(data.data(mime))
-        path = urlparse(url).path.strip()
-        scheme = urlparse(url).scheme
+        url = urlparse(str(data.data(mime), encoding='utf-8').rstrip())
```
What is the sample string, that `data.data(mime)` is returning?

I wonder if `.strip()` can be used instead of `.rstrip()` to be more resilient to unusual strings.

---

Changes in this file look like general bug fixes and should go into a standalone PR with an explanation of a problem (all inputs & outputs) and a solution.

stbenjam replied:
In the case of dropping a file, the path contains `\r\n` on the end, at least on Linux.  It looks like it breaks urlparser.  I don't see any added value to `.strip()`, since this is generated by the OS.  Why would it inject something at the beginning?

aik099 replied:
I see now. Thank you.

In fact, extra stripping code won't be necessary anymore, when you'll add a setting (on reader class-level) if they want file chooser dialog or a textarea with exported annotations.

aik099 commented on common_utils.py:
```diff
@@ -221,21 +221,24 @@ def dragMoveEvent(self, event):
-        url = str(data.data(mime))
-        path = urlparse(url).path.strip()
-        scheme = urlparse(url).scheme
+        url = urlparse(str(data.data(mime), encoding='utf-8').rstrip())
+        path = url.path.strip()
```
Is the `.strip()` needed here is we'll employ it on the `url` variable above?

aik099 replied:
I see this change was removed. I guess that's fair considering that now file dialog is being used and no extra line endings are added.

aik099 reviewed (changes requested):
I wonder if there isn't an easier way to get access to the Moon Reader database. Doesn't it store it somewhere on the mounted USB disk, as some Android apps do?

The https://www.mobileread.com/forums/showthread.php?t=328879&highlight=moon+reader+database thread suggests that it might be located under the `android/data` folder or similar.

stbenjam commented:
> I wonder if there isn't an easier way to get access to the Moon Reader database. Doesn't it store it somewhere on the mounted USB disk, as some Android apps do?

> The https://www.mobileread.com/forums/showthread.php?t=328879&highlight=moon+reader+database thread suggests that it might be located under the android/data folder or similar.

It appears to be in `/data/data`, which isn't accessible via USB.

aik099 commented:
> > I wonder if there isn't an easier way to get access to the Moon Reader database. Doesn't it store it somewhere on the mounted USB disk, as some Android apps do?
> 
> > The https://www.mobileread.com/forums/showthread.php?t=328879&highlight=moon+reader+database thread suggests that it might be located under the android/data folder or similar.
> 
> It appears to be in `/data/data`, which isn't accessible via USB.

I see. Then parsing backup is better than nothing for sure 😄 .

stbenjam committed:
Allow arbitrary HTML color, and setting default with picker

This introduces a default bg and fg color setting, and also allows
specifying an HTML color as well as the existing hard coded shades.

stbenjam committed:
Use file chooser

stbenjam committed:
Add all annotations, not just for selected book

stbenjam committed:
Add reader class help

stbenjam committed:
Set filechooser to existing file only

stbenjam committed:
Fix window title and renamed pte var

stbenjam committed:
Only enable annotations that have any confidence by default

stbenjam committed:
Update help

stbenjam committed:
Remove stray reference to Default color

stbenjam committed:
Consider title + author + filename match authoritative

stbenjam committed:
Update help

stbenjam committed:
Fix renamed vars

stbenjam committed:
Adjust help

stbenjam committed:
Remove my debug statements

stbenjam committed:
Revert "Consider title + author + filename match authoritative"

This is causing Calibre to crash on my production instance... but
not my dev environment.  Needs investigating

This reverts commit 393fd770488724753a43236572189c23f2d0cbd0.

aik099 commented on action.py:
```diff
@@ -23,11 +23,11 @@
 
 try:
-    from PyQt5.Qt import (pyqtSignal, Qt, QApplication, QIcon, QMenu, QPixmap,
+    from PyQt5.Qt import (pyqtSignal, Qt, QApplication, QIcon, QMenu, QMessageBox, QPixmap,
```
Please remove code, that is now extracted into separate PR's like #25 and #24.

stbenjam replied:
This will land last and I'll rebase, I don't need to remove it here. I'd wait to review this PR until those others merge.

aik099 replied:
OK.

aik099 commented on action.py:
```diff
@@ -95,8 +95,8 @@ class AnnotationsAction(InterfaceAction, Logger):
     SELECT_DESTINATION_DET_MSG = _(
-        "To determine which book will receive incoming annotations, annotation metadata (Title, Author, UUID) is compared to library metadata.\n\n"
-        "Annotations whose metadata completely matches library metadata will be added automatically to the corresponding book.\n\n"
+        "To determine which book will receive incoming annotations, annotation metadata (Title, Author, UUID, Filename) is compared to library metadata.\n\n"
+        "Annotations whose metadata matches a title, author and UUID or Filename in the Library metadata will be added automatically to the corresponding book.\n\n"
```
Looks like another unrelated change, that deserves extracting.

P.S.
I love strictly on topic PRs :)

aik099 commented on action.py:
```diff
@@ -808,23 +807,29 @@ def import_annotations(self, reader_app):
         self.selected_mi = get_selected_book_mi(self.get_options(),
                                                 msg=self.SELECT_DESTINATION_MSG,
                                                 det_msg=self.SELECT_DESTINATION_DET_MSG)
-        if not self.selected_mi:
-            return
```
Why this was removed? What it actually does?

Not a part of #25 (not sure if it should be).

aik099 commented on annotated_books.py:
```diff
@@ -251,6 +249,9 @@ def __init__(self, parent, book_list, get_annotations_as_HTML, source):
                 if confidence is None:
                     cid, confidence = parent.generate_confidence(book_data)
 
+            enabled = QCheckBox()
+            enabled.setChecked(confidence)
+
```
Several problems here:

1. unrelated change (or even feature) > move to another PR with an explanation and benefits
2. the `confidence` variable has a numerical value (from `0` to `5`), but the `enabled.setChecked(...)` method accepts a boolean value, which might lead to a type error > suggesting to replace `confidence` with `confidence > 0` (hopefully that will yield a boolean value as a result)

aik099 commented on annotations.py:
```diff
@@ -158,6 +162,7 @@ def to_HTML(self, header=''):
 #         self._log_location("comments_body='%s'" % comments_body)
 
         if self.annotations:
+            html_color_pattern = re.compile("^#(?:[0-9a-fA-F]{3,4}){1,2}$")
```
Please remove code (across this PR), that is now extracted into separate PRs like #24.

stbenjam replied:
This will land last and I'll rebase, I don't need to remove it here. I'd wait to review this PR until those others merge.

aik099 replied:
OK.

stbenjam committed:
Add specific `None` case back.

stbenjam commented:
I moved the title to WIP since this isn't ready to review. After the initial set of comments, I realized I should break out some additional features into additional PR's.  Since this is dependent on those, I would wait to look at anything here until I rebase this one after those land.

stbenjam committed:
Fix syncing??

stbenjam committed:
Hack for multi syncing?

aik099 commented:
> I moved the title to WIP since this isn't ready to review. After the initial set of comments, I realized I should break out some additional features into additional PR's. Since this is dependent on those, I would wait to look at anything here until I rebase this one after those land.

You can convert a PR into a draft. More info about this at https://github.blog/changelog/2020-04-08-convert-pull-request-to-draft/ .

stbenjam commented:
Sure, I made it a draft if that's somehow better for you than [WIP] title :shrug:

stbenjam commented:
It's been a month, this repo despite claims to the contrary doesn't seem to be *actively* maintained with just a few commits per year, and little activity on my forum questions.

aik099 commented:
@stbenjam , I completely understand you. Please be patient. We're not in a hurry here with @davidfor being very occupied with other projects and not being able to timely review any PR coming in.

I'm glad, that you're able to implement a Moon+ Reader integration, that works for you. Please don't let it disappear and not make any other Moon+ Reader + Calibre user happy.

It will be reviewed eventually I guess.

stbenjam commented:
Nah, I'm good.  I'm busy too.

Geo123abc commented:
@stbenjam   I'm not good at coding. And I have not got your branch on time. If convenient, can you share the new calibre-annotations plugin that you show in this page which support Moon+ Reader？I need it very much . Thanks a lot.

davidfor commented:
@stbenjam: This repository is maintained at exactly the rate it is needed to. Yes, there are few commits per year. But, that is because there are very few changes that have been needed. 

For your change, you had perfectly bad timing. Some work and personal commitments hit me at that time. That left no time and no energy for anything else. Those extended for longer than expected and things that are unimportant to me, which unfortunately includes this plugin, went by the wayside. When I did find the time, and I admit, remembered it needed to be done, I discovered the pull request had been closed a few days before.

That was a couple of weeks ago, and I should have posted this then. But, honestly, if you have that little interest that you actually withdrew the pull request, I figured my time was better spent doing something else. Such as reading a book.

If you do want this done, please open another pull request. I will try an look at it over the weekend after you do that.

dairefagan commented:
@stbenjam I am very interested to try your plugin as it would form a key component in a workflow concept. I hope to see it implemented soon.

stbenjam commented:
The branches are restored if someone wants to pick up working on them. 

I'm sympathetic to being busy, but I also have plenty of experience in open source and attracting contributors requires at least some acknowledgement and communication about what to expect.   I can't go months between reviews, given the loss of context between each round. No hard feelings if you can't keep up on what needs to be done in the project, but I'm incredulous to call it a maintained project when there's multiple PR's still open on this repo without a single comment.

davidfor commented:
You said, "doesn't seem to be actively maintained with just a few commits per year". That was what I was referring to. The number of commits each year is meaningless because there were no changes needed. Am I supposed to make up things to do? 

I  suppose that I should have posted on these PRs something. But, when they came in, I planned to look at them in a few days, probably over the next weekend. If that had happened, everything would have gone through with no issues. Unfortunately, other things came up which meant I was not able to and also didn't think to post anything else.

In any case, my intention is to look at these this weekend. Hopefully no one breaks anything at work.

dairefagan commented:
Great to see this progressing. Thanks both 👍

davidfor commented:
I have finally found time to read through this and have a lot of problems with it. There are changes in it that have nothing to do with adding Moon+ Reader support. Some of that is in a pull request that has been merged. But, some are changes that not needed for this and may cause issues elsewhere. There is a comment about rebasing after the pull requests for the file chooser and colour support is done. These have been merged. Could you update this and resubmit?
